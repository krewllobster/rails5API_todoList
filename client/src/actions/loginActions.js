export const FETCH_LOGIN_REQUEST = 'FETCH_LOGIN_REQUEST'
export const FETCH_LOGIN_SUCCESS = 'FETCH_LOGIN_SUCCESS'
export const FETCH_LOGIN_FAILURE = 'FETCH_LOGIN_FAILURE'
export const LOGOUT = 'LOGOUT'

export const logout = () => ({
  type: LOGOUT,
})

export const fetchLoginRequest = () => ({
  type: FETCH_LOGIN_REQUEST,
})

export const fetchLoginSuccess = (apiKey) => ({
  type: FETCH_LOGIN_SUCCESS,
  apiKey,
})

export const fetchLoginFailure = (authErrorMessage) => ({
  type: FETCH_LOGIN_FAILURE,
  authErrorMessage,
})

export const fetchLogin = (urlQueryParams) => {
  return (dispatch) => {
    dispatch(fetchLoginRequest())

    return fetch(`auth/login?${urlQueryParams}`, {
      method: 'post',
    })
    .then(
      response => response.json()
    )
    .then(json => {
      if (json.auth_token) {
        dispatch(fetchLoginSuccess(json.auth_token))
      } else {
        dispatch(fetchLoginFailure(json.message))
      }
    })
  }
}
