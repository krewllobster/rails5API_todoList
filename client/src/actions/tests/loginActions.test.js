import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import fetchMock from 'fetch-mock'
import * as actions from '../loginActions'
import {
  FETCH_LOGIN_REQUEST,
  LOGOUT,
  FETCH_LOGIN_FAILURE,
  FETCH_LOGIN_SUCCESS,
} from '../loginActions'

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

const mockResultSuccess = {
  auth_token: 'asdf'
}

const mockResultFailure = {
  message: 'boo'
}

const urlQueryParams = 'email=test@test.com&password=test'

describe('async-actions', () => {
  it('calls login request and success action if response was successful', () => {

    fetchMock.postOnce(
      `auth/login?${urlQueryParams}`,
      mockResultSuccess
    )

    const expectedActions = [
      { type: FETCH_LOGIN_REQUEST },
      { type: FETCH_LOGIN_SUCCESS, apiKey: 'asdf' }
    ]

    const store = mockStore({ auth: { apiKey: null} })

    return store.dispatch(actions.fetchLogin(urlQueryParams))
      .then((data) => {
        expect(store.getActions()).toEqual(expectedActions)
      })
  })

  it('calls login request and failure action if response was unsuccessful', () => {

    fetchMock.postOnce(
      `auth/login?${urlQueryParams}`,
      mockResultFailure
    )

    const expectedActions = [
      { type: FETCH_LOGIN_REQUEST },
      { type: FETCH_LOGIN_FAILURE, authErrorMessage: 'boo'}
    ]

    const store = mockStore({ auth: { authErrorMessage: ''} })

    return store.dispatch(actions.fetchLogin(urlQueryParams))
      .then((data) => {
        expect(store.getActions()).toEqual(expectedActions)
      })

  })
})

describe('actions', () => {
  const store=mockStore({})
  expect(store.dispatch(actions.fetchLoginRequest())).toMatchSnapshot()
  expect(store.dispatch(actions.fetchLoginSuccess('fake_auth_token'))).toMatchSnapshot()
  expect(store.dispatch(actions.fetchLoginFailure('fake_auth_error_message'))).toMatchSnapshot()
  expect(store.dispatch(actions.logout())).toMatchSnapshot()
})
