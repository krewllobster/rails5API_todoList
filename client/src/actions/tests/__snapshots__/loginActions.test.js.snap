// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "type": "FETCH_LOGIN_REQUEST",
}
`;

exports[` 2`] = `
Object {
  "apiKey": "fake_auth_token",
  "type": "FETCH_LOGIN_SUCCESS",
}
`;

exports[` 3`] = `
Object {
  "authErrorMessage": "fake_auth_error_message",
  "type": "FETCH_LOGIN_FAILURE",
}
`;

exports[` 4`] = `
Object {
  "type": "LOGOUT",
}
`;
